Makefile

```diff
@@ -48,7 +50,21 @@ else
 	else
 		VERSION ?= master
 	endif
-	GITEA_VERSION ?= $(shell git describe --tags --always | sed 's/-/+/' | sed 's/^v//')
+
+	STORED_VERSION=$(shell cat $(STORED_VERSION_FILE) 2>/dev/null)
+	ifneq ($(STORED_VERSION),)
+		GITEA_VERSION ?= $(STORED_VERSION)
+	else
+		GITEA_VERSION ?= $(shell git describe --tags --always | sed 's/-/+/' | sed 's/^v//')
+	endif
+endif
+
+ifeq (,$(filter build-offline,$(MAKECMDGOALS)))
+	NODE_MODULES := node_modules
+	NPX := npx
+else
+	NODE_MODULES :=
```

Comment:
Can you elaborate? I think the concept of storing/restoring the version and the offline build are two distinct concepts that should probably not be tightly coupled.

Reply:
```
if online build or version-file inexisting
   version from git
else
   version from file
```

The idea is version file acts as the offline tool, while git version should be used for online builds.

Reply:
imho it would be nice to use version independently of online/offline build. Version file is more needed if there is no git folder and that could be possibility also for online build

Reply:
Whatever it's online or offline build, we can both use

if version-file inexisting
   version from git
else
   version from file

Reply:
@lunny I think this is how this already works: https://github.com/go-gitea/gitea/pull/10325/files/b6e5cb837a66dcdfb87a0a027f6dc038db580020#diff-b67911656ef5d18c4ae36cb6741b7965R54